File: android/ClipSyncAndroid/app/src/main/java/com/clipsync/android/crypto/SasCode.kt
package com.clipsync.android.crypto

/**
 * Short Authentication String — the 6-digit pairing code, matching the Mac
 * `SASCode`. Both devices sort the two raw public keys, concatenate, then
 * HKDF-SHA256 (info "clipsync-sas") to 4 bytes read big-endian, mod 1_000_000.
 */
object SasCode {

    fun derive(keyA: ByteArray, keyB: ByteArray): String {
        val sorted = if (lexicographicallyPrecedes(keyA, keyB)) keyA + keyB else keyB + keyA
        val derived = Hkdf.derive(
            ikm = sorted,
            salt = ByteArray(0),
            info = "clipsync-sas".toByteArray(Charsets.UTF_8),
            length = 4,
        )
        var value = 0L
        for (b in derived) value = (value shl 8) or (b.toLong() and 0xff)
        return (value % 1_000_000L).toString().padStart(6, '0')
    }

    private fun lexicographicallyPrecedes(a: ByteArray, b: ByteArray): Boolean {
        val n = minOf(a.size, b.size)
        for (i in 0 until n) {
            val ai = a[i].toInt() and 0xff
            val bi = b[i].toInt() and 0xff
            if (ai != bi) return ai < bi
        }
        return a.size < b.size
    }
}

File: mac/ClipSyncCore/Sources/ClipSyncCore/SASCode.swift
import Foundation
import CryptoKit

/// Short Authentication String — the 6-digit code both devices derive from their
/// two public keys and show to the user during first-time pairing
/// (see `docs/pairing-flow.md`). If the codes match, there is no
/// man-in-the-middle and the keys are authentic.
///
/// Both devices MUST derive the identical code, so the byte encoding below is the
/// cross-platform contract with the Android app:
///
/// 1. Sort the two raw 32-byte public keys by their bytes (so order doesn't
///    matter) and concatenate → a 64-byte transcript.
/// 2. `HKDF-SHA256(ikm: transcript, salt: empty, info: "clipsync-sas", L: 4)`.
/// 3. Read the 4 output bytes as a big-endian `UInt32`, take it `mod 1_000_000`,
///    zero-pad to 6 digits.
public enum SASCode {
    /// HKDF `info` binding the derivation to the SAS purpose.
    static let info = Data("clipsync-sas".utf8)

    /// Derives the 6-digit code from two raw 32-byte X25519 public keys.
    /// Order-independent by construction.
    public static func derive(_ keyA: Data, _ keyB: Data) -> String {
        let sorted = keyA.lexicographicallyPrecedes(keyB) ? [keyA, keyB] : [keyB, keyA]
        var transcript = Data()
        transcript.append(sorted[0])
        transcript.append(sorted[1])

        let derived = HKDF<SHA256>.deriveKey(
            inputKeyMaterial: SymmetricKey(data: transcript),
            salt: Data(),
            info: info,
            outputByteCount: 4
        )
        let bytes = derived.withUnsafeBytes { Data($0) }
        let value = bytes.reduce(UInt32(0)) { ($0 << 8) | UInt32($1) }
        return String(format: "%06u", value % 1_000_000)
    }
}
